server/core/db/boltdb.go

```diff
@@ -101,28 +109,90 @@ func (b *BoltDB) TryLock(newLock models.ProjectLock) (bool, models.ProjectLock,
 			currLock = newLock
 			return nil
 		}
+		// checking if current lock is with the same PR
+		var currLock models.ProjectLock
+		if err := json.Unmarshal(currLockSerialized, &currLock); err != nil {
+			return errors.Wrap(err, "failed to deserialize current lock")
+		}
+		if currLock.Pull.Num == newLock.Pull.Num {
+			lockAcquired = true
+			return nil
+		}
 
 		// otherwise the lock fails, return to caller the run that's holding the lock
 		if err := json.Unmarshal(currLockSerialized, &currLock); err != nil {
 			return errors.Wrap(err, "failed to deserialize current lock")
 		}
 		lockAcquired = false
+
+		// Enqueuing
+		queueBucket := tx.Bucket(b.queueBucketName)
+		currQueueSerialized := queueBucket.Get([]byte(key))
+		// Queue doesn't exist, create one
+		if currQueueSerialized == nil {
+			newQueue := []models.ProjectLock{newLock}
+			newQueueSerialized, _ := json.Marshal(newQueue)
+			if err := queueBucket.Put([]byte(key), newQueueSerialized); err != nil {
+				return err
+			}
+			enqueueStatus = models.EnqueueStatus{
+				Status:              models.Enqueued,
+				ProjectLocksInFront: 1,
+			}
+			return nil
+		}
+		// Queue exists
+		var currQueue []models.ProjectLock
+		if err := json.Unmarshal(currQueueSerialized, &currQueue); err != nil {
+			return errors.Wrap(err, "failed to deserialize queue for current lock")
+		}
+
+		// Lock is already in the queue
+		if indexInQueue := findPullRequest(currQueue, newLock.Pull.Num); indexInQueue > -1 {
+			enqueueStatus = models.EnqueueStatus{
+				Status:              models.AlreadyInTheQueue,
+				ProjectLocksInFront: indexInQueue + 1,
+			}
+			return nil
+		}
+		// Not in the queue, add it
+		newQueue := append(currQueue, newLock)
+		newQueueSerialized, _ := json.Marshal(newQueue)
+		if err := queueBucket.Put([]byte(key), newQueueSerialized); err != nil {
+			return err
+		}
+		enqueueStatus = models.EnqueueStatus{
+			Status:              models.Enqueued,
+			ProjectLocksInFront: len(newQueue),
+		}
+
 		return nil
 	})
 
 	if transactionErr != nil {
-		return false, currLock, errors.Wrap(transactionErr, "DB transaction failed")
+		return false, currLock, enqueueStatus, errors.Wrap(transactionErr, "DB transaction failed")
+	}
+
+	return lockAcquired, currLock, enqueueStatus, nil
+}
+
+func findPullRequest(locks []models.ProjectLock, pullRequestNumber int) int {
+	for i := range locks {
+		if locks[i].Pull.Num == pullRequestNumber {
+			return i
+		}
 	}
 
-	return lockAcquired, currLock, nil
+	return -1
 }
 
 // Unlock attempts to unlock the project and workspace.
 // If there is no lock, then it will return a nil pointer.
 // If there is a lock, then it will delete it, and then return a pointer
 // to the deleted lock.
```

Comment:
```suggestion
// Unlock attempts to unlock the project and workspace, and if the queue is not empty, it will pass on the lock to the next item in the queue
// If there is no lock, then no changes will be made, and the method will return two nil pointers.
// If there is a lock, then the method will delete it. On top of it, if the queue is not empty, it will pass on the lock to the next item in the queue. Next, it will return pointers to:
// - the deleted lock
// - the dequeued lock (if any)
```

Reply:
Documentation updated in https://github.com/runatlantis/atlantis/pull/2782